src/Provider/AbstractProvider.php

```diff
@@ -98,7 +108,15 @@ public function __construct($options = [], HttpAdapterInterface $httpClient = nu
             }
         }
 
-        $this->setHttpClient($httpClient ?: new CurlHttpAdapter());
+        if (empty($collaborators['httpClient'])) {
+            $collaborators['httpClient'] = new CurlHttpAdapter();
+        }
+        $this->setHttpClient($collaborators['httpClient']);
+
+        if (empty($collaborators['randomFactory'])) {
+            $collaborators['randomFactory'] = new RandomFactory();
+        }
+        $this->setRandomFactory($collaborators['randomFactory']);
```

Comment:
If we use the factory, then any library other than RandomLib will need to implement a factory that extends RandomLib's factory, or the user will need to do that on their own. My point is: if we only require that they pass in a Generator, then the user may simply wrap whatever other library they are using to adhere to the Generator interface and pass it in, without needing to also implement a factory.